protocol.md

```diff
@@ -726,6 +727,141 @@ Upload-Length: 11
+
+Where `Upload-Offset` and `Content-Length` headers are used in the same request, their
+values MUST be used as the respective `Upload-Offset` and `Content-Length` values. A
+value of `0` MUST be used for `Upload-Offset` and `Content-Length` where no headers
+of the same name are sent with the request.
+
+Servers receiving a `HEAD` or `PATCH` request with a missing or invalid `Upload-Challenge` value
+MUST respond with a `404 Not Found` status.
+
+For requests that reference multiple upload resources, the `Upload-Challenge` is computed by
+concatenating the `Upload-Challenge`s of the individual upload resources for the request in the
+order they are referenced. The SHA256 checksum computed over the concatenated `Upload-Challenge`s
+is then used as the `Upload-Challenge` for the request:
+
+```
+Upload-Challenge = SHA256([Upload-Challenge for resource 1] + … + [Upload-Challenge for resource n])
+```
+
+#### Examples
+
+An upload resource is created with [Creation With Upload](#creation-with-upload) and an `Upload-Secret`:
+
+```
+POST /files HTTP/1.1
+Upload-Length: 100
+Upload-Secret: R290Y2hhISBVc2UgYW4gYWN0dWFsIGhpZ2gtZW50cm9weSBzb3VyY2UgaW4gcmVhbCBjb2RlIDspCg
+Content-Length: 5
+Content-Type: application/offset+octet-stream
+Tus-Resumable: 1.0.0
+
+hello
+```
+
+The upload is continued with a `PATCH` request, including an `Upload-Challenge` computed as
+
+```
+Upload-Challenge = SHA256("R290Y2hhISBVc2UgYW4gYWN0dWFsIGhpZ2gtZW50cm9weSBzb3VyY2UgaW4gcmVhbCBjb2RlIDspCg" + SHA256("PATCH" + "5" + "R290Y2hhISBVc2UgYW4gYWN0dWFsIGhpZ2gtZW50cm9weSBzb3VyY2UgaW4gcmVhbCBjb2RlIDspCg" + "50"))
+```
+
+and included in the request:
+
+```
+PATCH /files/24e533e02ec3bc40c387f1a0e460e216 HTTP/1.1
+Upload-Offset: 5
+Upload-Challenge: ed9e63743b8073e97a1bd54d0f92161fa32c27bff130f11ff896e6e0271f2688
+Content-Length: 50
+Content-Type: application/offset+octet-stream
+Tus-Resumable: 1.0.0
+
+[50 bytes of content]
+```
+
+Then, a `DELETE` request for the upload is sent, including an `Upload-Challenge` computed as
+
+```
+Upload-Challenge = SHA256("R290Y2hhISBVc2UgYW4gYWN0dWFsIGhpZ2gtZW50cm9weSBzb3VyY2UgaW4gcmVhbCBjb2RlIDspCg" + SHA256("DELETE" + "0" + "R290Y2hhISBVc2UgYW4gYWN0dWFsIGhpZ2gtZW50cm9weSBzb3VyY2UgaW4gcmVhbCBjb2RlIDspCg" + "0"))
+```
+
+and included in the request:
+
+```
+DELETE /files/24e533e02ec3bc40c387f1a0e460e216 HTTP/1.1
+Content-Length: 0
+Upload-Challenge: 59ca949f7eed4c4a1f7e40da7c119fb7cb1095262a12eab770685711ea5a176a
+Tus-Resumable: 1.0.0
+
+[50 bytes of content]
+```
+
+Two uploads - each with its own `Upload-Secret` - are concatenated, including an `Upload-Challenge`. For `/files/a`, the `Upload-Challenge` is computed as:
+
+```
+Upload-Challenge-1 = SHA256("VGhpcyBpc24ndCBoaWdoLWVudHJvcHkgZWl0aGVyLiBEb24ndCB0cnkgdGhpcyBhdCBob21lLCBraWRzIQ" + SHA256("POST" + "0" + "VGhpcyBpc24ndCBoaWdoLWVudHJvcHkgZWl0aGVyLiBEb24ndCB0cnkgdGhpcyBhdCBob21lLCBraWRzIQ" + "0"))
+Upload-Challenge-1 = 1824949792ae9040811f1888f992608de8da7c17f6f6f5bc966dc5c95a1acba3
```

Comment:
Should the "Upload-Offset" be calculated as zero hero or should we force the actual Upload-Offset value? The file should be completed already for it to be usable for concatenation so the client would already know the offset.

Reply:
I like the idea, but I'm not sure it will work in all circumstances. For concatenation, the actual `Upload-Offset` should be known, but in the future there could be other use cases where it is not known.

Also, the `Upload-Offset` and `Content-Length` parts in the computation are currently defined as having to match the values submitted in the request's header - which would include concatenation requests as well. That definition would need to be adapted as well, then, adding complexity.